fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepository.java

```diff
@@ -16,7 +16,228 @@
+import com.google.common.collect.Table;
+import com.netflix.spinnaker.fiat.model.UserPermission;
+import com.netflix.spinnaker.fiat.model.resources.Account;
+import com.netflix.spinnaker.fiat.model.resources.Application;
+import com.netflix.spinnaker.fiat.model.resources.Named;
+import com.netflix.spinnaker.fiat.model.resources.Resource;
+import com.netflix.spinnaker.fiat.redis.JedisSource;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.Pipeline;
+import redis.clients.jedis.Response;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * This Redis-backed permission repository is structured in a way to optimized reading types of
+ * resource permissions. In general, this looks like a key schema like:
+ *
+ * "prefix:myuser@domain.org:resources" -> {"resourceName1" -> "<serialized json of resourceName1>",
+ *                                          "resourceName2" -> "<serialized json of resourceName1>"}
+ *
+ * Additionally, a helper key, called the "all users" key, maintains a set of all usernames.
+ */
+@Component
+@Slf4j
+public class RedisPermissionsRepository implements PermissionsRepository {
+
+  private static final String KEY_PERMISSIONS = "permissions";
+  private static final String KEY_ALL_USERS = "users";
+
+  @Value("${fiat.redis.prefix:spinnaker:fiat}")
+  @Setter
+  private String prefix;
+
+  @Autowired
+  @Setter
+  private ObjectMapper objectMapper;
+
+  @Autowired
+  @Setter
+  private JedisSource jedisSource;
+
+  // TODO(ttomsu): Add RedisCacheOptions from Clouddriver.
+
+  @Override
+  public PermissionsRepository put(@NonNull UserPermission permission) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      for (Resource r : Resource.values()) {
+        Map<String, String> resourceValues = new HashMap<>();
+
+        Consumer<Named> putInValuesMap = namedResource -> {
+          try {
+            resourceValues.put(namedResource.getName(),
+                               objectMapper.writeValueAsString(namedResource));
+          } catch (JsonProcessingException jpe) {
+            log.error("Serialization exception writing " + permission.getId() + " entry.", jpe);
+          }
+        };
+
+        switch (r) {
+          case ACCOUNT:
+            permission.getAccounts().stream().forEach(putInValuesMap);
+            break;
+          case APPLICATION:
+            permission.getApplications().stream().forEach(putInValuesMap);
+            break;
+        }
+
+        if (!resourceValues.isEmpty()) {
+          jedis.hmset(userKey(permission.getId(), r), resourceValues);
```

Comment:
could theoretically do `.pipelined()` before the `for (Resource r`, and do these calls on the pipeline, then commit it at after the end of the for loop